oscore-cf/src/main/java/org/eclipse/californium/oscore/OSCoreClient.java

```diff
@@ -0,0 +1,104 @@
+ * and Eclipse Distribution License v1.0 which accompany this distribution.
+ * 
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ * 
+ * Contributors:
+ *    Joakim Brorsson
+ *    Tobias Andersson (RISE SICS)
+ *    
+ ******************************************************************************/
+package org.eclipse.californium.oscore;
+
+import org.eclipse.californium.core.CoapClient;
+import org.eclipse.californium.core.coap.BlockOption;
+import org.eclipse.californium.core.coap.Option;
+import org.eclipse.californium.core.coap.OptionNumberRegistry;
+import org.eclipse.californium.core.coap.Request;
+import org.eclipse.californium.core.network.Endpoint;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.Collections;
+
+public class OSCoreClient extends CoapClient {
+    private static OSCoreEndpoint newEndpoint() {
+        OSCoreEndpoint ep = new OSCoreEndpoint(Collections.<String>emptyList());
+        try {
+            ep.start();
+        } catch (IOException e) {
+            System.out.println("Could not create endpoint: " + e.getMessage());
+        }
+        return ep;
+    }
+
+    /**
+     * Constructs a new CoapClient that has no destination URI yet.
+     * 
+     * @param db  the OSCore context database
+     */
+    public OSCoreClient(OSCoreCtx ctx) {
+        super();
+        setEndpoint(newEndpoint());
+    }
+
+    /**
+     * Constructs a new CoapClient that sends requests to the specified URI.
+     *
+     * @param uri the uri
+     * @param db  the OSCore context database
+     */
+    public OSCoreClient(String uri) {
+        super(uri);
+        setEndpoint(newEndpoint());
+    }
+
+    /**
+     * Constructs a new CoapClient that sends request to the specified URI.
+     *
+     * @param uri the uri
+     * @param db  the OSCore context database
+     */
+    public OSCoreClient(URI uri) {
+        super(uri);
+        setEndpoint(newEndpoint());
+    }
+
+    /**
+     * Constructs a new CoapClient with the specified scheme, host, port and
+     * path as URI.
+     * 
+     * @param db the OSCore context database     *
+     * @param scheme the scheme
+     * @param host the host
+     * @param port the port
+     * @param path the path
+     */
+    public OSCoreClient(String scheme, String host, int port, String... path) {
+        super(scheme, host, port, path);
+        setEndpoint(newEndpoint());
+    }
+
+
+    /**
+     * Sends the specified request over the specified endpoint.
+     *
+     * @param request the request
+     * @param outEndpoint the endpoint
+     * @return the request
+     */
+    @Override
+    protected Request send(Request request, Endpoint outEndpoint) {
+    	OSCoreEndpoint ep = (OSCoreEndpoint) outEndpoint;
+    	request.getOptions().setOscore(new byte[0]);
```

Comment:
Please reuse the empty `byte[0]`.

Reply:
Why? I just want to initiate an empty Oscore option and in this case is an empty byte array.

Reply:
Because you frequently allocate a new immutable byte[0] instance.
If you use a `static final byte[] EMPTY = new byte[0];` you don't stress the GC.

Reply:
Thanks! Now I understand.